pkg/controllers/persistentvolumeclaim/controller.go

```diff
@@ -0,0 +1,126 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistentvolumeclaim
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/aws/karpenter/pkg/utils/functional"
+	"github.com/aws/karpenter/pkg/utils/injection"
+	"github.com/aws/karpenter/pkg/utils/pod"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"knative.dev/pkg/logging"
+
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName         = "volume"
+	SelectedNodeAnnotation = "volume.kubernetes.io/selected-node"
+)
+
+// Controller for the resource
+type Controller struct {
+	kubeClient client.Client
+}
+
+// NewController is a constructor
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{kubeClient: kubeClient}
+}
+
+// Register the controller to the manager
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	return controllerruntime.
+		NewControllerManagedBy(m).
+		Named(controllerName).
+		For(&v1.PersistentVolumeClaim{}).
+		Watches(&source.Kind{Type: &v1.Pod{}}, handler.EnqueueRequestsFromMapFunc(c.pvcForPod)).
+		Complete(c)
+}
+
+// Reconcile a control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named(controllerName).With("resource", req.String()))
+	ctx = injection.WithNamespacedName(ctx, req.NamespacedName)
+	ctx = injection.WithControllerName(ctx, controllerName)
+
+	pvc := &v1.PersistentVolumeClaim{}
+	if err := c.kubeClient.Get(ctx, req.NamespacedName, pvc); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	pod, err := c.podForPvc(ctx, pvc)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if pod == nil {
+		return reconcile.Result{}, nil
+	}
+	if nodeName, ok := pvc.Annotations[SelectedNodeAnnotation]; ok && nodeName == pod.Spec.NodeName {
+		return reconcile.Result{}, nil
+	}
+	if !c.isBindable(pod) {
+		return reconcile.Result{}, nil
+	}
+	pvc.Annotations = functional.UnionStringMaps(pvc.Annotations, map[string]string{SelectedNodeAnnotation: pod.Spec.NodeName})
+	if err := c.kubeClient.Update(ctx, pvc); err != nil {
```

Comment:
I think I'm missing something obvious but why is this necessary? Once we bind the pod with the PVC volume to that node, won't this happen anyway?

Reply:
No! I thought that the PVC controller should be responsible for this, but it turns out that it's a prebind implementation within the kube scheduler. Any scheduler implementation (i.e. karpenter) is responsible for also doing volume binding. 

This is unintuitive to me, and I feel like the kvc or pvc controller (or kubescheduler, for that matter) should be responsible for this. This was brought up by @wongma7 in a previous comment.

Reply:
TIL!